File: architecture-skills/design-patterns/examples/command/command.cpp
// Command pattern in C++.
//
// Text editor with undo: a virtual Command interface with execute/undo,
// concrete commands capturing their own undo state, and a history stack
// of unique_ptrs as the invoker. For one-shot commands without undo,
// std::function<void()> is the lighter modern idiom.

#include <iostream>
#include <memory>
#include <string>
#include <vector>

struct Editor {
    std::string text; // receiver
};

class Command {
public:
    virtual ~Command() = default;
    virtual void execute() = 0;
    virtual void undo() = 0;
};

class InsertText : public Command {
public:
    InsertText(Editor& editor, std::string text)
        : editor_(editor), text_(std::move(text)) {}

    void execute() override { editor_.text += text_; }
    void undo() override { editor_.text.erase(editor_.text.size() - text_.size()); }

private:
    Editor& editor_;
    std::string text_;
};

class DeleteLast : public Command {
public:
    DeleteLast(Editor& editor, std::size_t count) : editor_(editor), count_(count) {}

    void execute() override {
        deleted_ = editor_.text.substr(editor_.text.size() - count_);
        editor_.text.erase(editor_.text.size() - count_);
    }
    void undo() override { editor_.text += deleted_; }

private:
    Editor& editor_;
    std::size_t count_;
    std::string deleted_; // captured on execute so undo can restore it
};

class History {
public:
    void execute(std::unique_ptr<Command> command) {
        command->execute();
        undoStack_.push_back(std::move(command));
    }

    void undo() {
        if (!undoStack_.empty()) {
            undoStack_.back()->undo();
            undoStack_.pop_back();
        }
    }

private:
    std::vector<std::unique_ptr<Command>> undoStack_;
};

int main() {
    Editor editor;
    History history;

    history.execute(std::make_unique<InsertText>(editor, "Hello"));
    history.execute(std::make_unique<InsertText>(editor, ", world!"));
    history.execute(std::make_unique<DeleteLast>(editor, 1));
    std::cout << "after edits: \"" << editor.text << "\"\n";

    history.undo();
    history.undo();
    std::cout << "after 2 undos: \"" << editor.text << "\"\n";
    return 0;
}

File: architecture-skills/design-patterns/examples/command/command.rs
// Command pattern in Rust.
//
// Text editor with undo: a Command trait with execute/undo, boxed
// commands in a history vector. Commands borrow nothing — they receive
// the editor at call time — which keeps the borrow checker happy. For
// one-shot commands without undo, Box<dyn FnOnce(&mut Editor)> suffices.

struct Editor {
    text: String, // receiver
}

trait Command {
    fn execute(&mut self, editor: &mut Editor);
    fn undo(&mut self, editor: &mut Editor);
}

struct InsertText {
    text: String,
}

impl Command for InsertText {
    fn execute(&mut self, editor: &mut Editor) {
        editor.text.push_str(&self.text);
    }

    fn undo(&mut self, editor: &mut Editor) {
        let new_len = editor.text.len() - self.text.len();
        editor.text.truncate(new_len);
    }
}

struct DeleteLast {
    count: usize,
    deleted: String, // captured on execute so undo can restore it
}

impl Command for DeleteLast {
    fn execute(&mut self, editor: &mut Editor) {
        let split = editor.text.len() - self.count;
        self.deleted = editor.text.split_off(split);
    }

    fn undo(&mut self, editor: &mut Editor) {
        editor.text.push_str(&self.deleted);
    }
}

struct History {
    undo_stack: Vec<Box<dyn Command>>,
}

impl History {
    fn execute(&mut self, mut command: Box<dyn Command>, editor: &mut Editor) {
        command.execute(editor);
        self.undo_stack.push(command);
    }

    fn undo(&mut self, editor: &mut Editor) {
        if let Some(mut command) = self.undo_stack.pop() {
            command.undo(editor);
        }
    }
}

fn main() {
    let mut editor = Editor { text: String::new() };
    let mut history = History { undo_stack: Vec::new() };

    history.execute(Box::new(InsertText { text: "Hello".into() }), &mut editor);
    history.execute(Box::new(InsertText { text: ", world!".into() }), &mut editor);
    history.execute(Box::new(DeleteLast { count: 1, deleted: String::new() }), &mut editor);
    println!("after edits: {:?}", editor.text);

    history.undo(&mut editor);
    history.undo(&mut editor);
    println!("after 2 undos: {:?}", editor.text);
}
